kbase-extension/static/kbase/js/common/jobManager.js

```diff
@@ -0,0 +1,281 @@
+     *
+     * @returns {object} the initialised job manager
+     */
+    function JobManager(config) {
+        const { model, bus, viewResultsFunction } = config;
+        if (!model || !bus || !viewResultsFunction) {
+            throw new Error(
+                'cannot initialise job manager widget without params "bus", "model", and "viewResultsFunction"'
+            );
+        }
+        let controlPanel = config.controlPanel || null;
+
+        function setControlPanel(panel) {
+            controlPanel = panel;
+        }
+
+        /**
+         * Update the execMessage panel with details of the active jobs
+         */
+        function updateToolbarJobStatus() {
+            if (!controlPanel) {
+                console.warn('controlPanel has not been initialised');
+                return;
+            }
+            controlPanel.setExecMessage(
+                Jobs.createCombinedJobState(model.getItem('exec.jobs.byStatus'))
+            );
+        }
+
+        /**
+         * Execute a function to view job results.
+         */
+        function viewResults() {
+            return viewResultsFunction ? viewResultsFunction() : null;
+        }
+
+        /* JOB MANAGEMENT */
+
+        /**
+         * Get the IDs of jobs with a certain status
+         *
+         * @param {array} statusList - array of statuses to find
+         * @param {array} validStates - array of valid statuses for this action (optional)
+         * @returns {array} job IDs to perform the action upon
+         */
+        function getJobIDsByStatus(statusList, validStates) {
+            if (validStates && validStates.length) {
+                const allInTheList = statusList.every((status) => {
+                    return validStates.includes(status);
+                });
+                if (!allInTheList) {
+                    console.error(
+                        `Invalid status supplied! Valid statuses: ${validStates.join(
+                            '; '
+                        )}; supplied: ${statusList.join('; ')}.`
+                    );
+                    return null;
+                }
+            }
+
+            const jobsByStatus = model.getItem('exec.jobs.byStatus');
+            return statusList
+                .map((status) => {
+                    return jobsByStatus[status] ? Object.keys(jobsByStatus[status]) : [];
+                })
+                .flat();
+        }
+
+        /**
+         *
+         * @param {string} action - what action is to occur (cancel or retry)
+         * @param {array} statusList - array of statuses that the action applies to
+         * @param {array} jobList - array of job IDs
+         * @returns {object} arguments to create a modal to confirm or cancel the action
+         */
+
+        function _generateDialogArgs(action, statusList, jobList) {
+            const statusIx = statusList.reduce((acc, curr) => {
```

Comment:
I think you could also use a Set here. Something like:
const statusSet = new Set(statusList)